src/HexagonOptimize.cpp

```diff
@@ -1118,6 +1190,221 @@ class OptimizePatterns : public IRMutator {
     }
 };
 
+class VectorReducePatterns : public IRMutator {
+    using IRMutator::visit;
+
+    // Check for interleaves of vectors with stride 1 like shuffle with indices:
+    // 0, 1, 2,..., window_size - 1,
+    // 1, 2, 3,..., window_size,
+    // 2, 3, 4,..., window_size + 1,
+    // .....
+    // window_size != lanes
+    // TODO: Their could be other patterns as well which we should match
+    static int is_stencil_interleave(const Expr &op, int window_size) {
+        int lanes = op.type().lanes();
+        internal_assert(lanes > window_size);
+        if (const Shuffle *shuff = op.as<Shuffle>()) {
+            for (int i = window_size; i < lanes; i++) {
+                if ((i % window_size != window_size - 1) &&
+                    (shuff->indices[i - window_size + 1] != shuff->indices[i])) {
+                    return false;
+                }
+            }
+            return true;
+        }
+        return false;
+    }
+
+    Expr visit(const Mul *op) override {
+        if (const VectorReduce *vr = op->a.as<VectorReduce>()) {
+            int in_lanes = vr->value.type().lanes();
+            int out_lanes = vr->type.lanes();
+            int rfac = in_lanes / out_lanes;
+            if (vr->op == VectorReduce::Add && rfac <= 4) {
+                // Try Moving broadcasts inside the VectorReduce Node for
+                // vrmpy/vtmpy/vdmpy instructions.
+                const int lanes = vr->type.lanes();
+                const int arg_lanes = vr->value.type().lanes();
+                IRMatcher::Wild<0> x;
+                IRMatcher::Wild<1> y;
+                auto rewrite = IRMatcher::rewriter(IRMatcher::mul(op->a, op->b), op->type);
+                if (rewrite(h_add(x, lanes) * broadcast(y, lanes), h_add(x * broadcast(y, arg_lanes), lanes))) {
+                    Expr new_expr = mutate(rewrite.result);
+                    if (!new_expr.same_as(rewrite.result)) {
+                        return new_expr;
+                    }
+                }
+            }
+        }
+        return IRMutator::visit(op);
+    }
+
+    typedef struct {
```

Comment:
Nit: `struct Signature { ... }`, also move this to be a local declaration in the visit function.